artifact/scripts/executor.cpp

```diff
@@ -0,0 +1,252 @@
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/path.hpp>
+
+
+namespace mender {
+namespace artifact {
+namespace scripts {
+namespace executor {
+
+namespace expected = mender::common::expected;
+
+
+using expected::ExpectedBool;
+
+namespace common = mender::common;
+namespace merror = mender::common::error;
+namespace path = mender::common::path;
+
+
+const int expected_state_script_version {3};
+
+unordered_map<const State, string> state_map {
+	{State::Idle, "Idle"},
+	{State::Sync, "Sync"},
+	{State::Download, "Download"},
+	{State::ArtifactInstall, "ArtifactInstall"},
+	{State::ArtifactReboot, "ArtifactReboot"},
+	{State::ArtifactCommit, "ArtifactCommit"},
+	{State::ArtifactRollback, "ArtifactRollback"},
+	{State::ArtifactRollbackReboot, "ArtifactRollbackReboot"},
+	{State::ArtifactFailure, "ArtifactFailure"},
+};
+
+unordered_map<const Action, string> action_map {
+	{Action::Enter, "Enter"},
+	{Action::Leave, "Leave"},
+	{Action::Error, "Error"},
+};
+
+merror::Error CorrectVersionFile(const string &path) {
+	ifstream vf {path};
+
+	// Missing file is OK
+	if (!vf.is_open()) {
+		return merror::NoError;
+	}
+
+	if (!vf) {
+		auto errnum {errno};
+		return merror::Error(
+			generic_category().default_error_condition(errnum), "Failed to open the version file");
+	}
+
+	string version;
+	vf >> version;
+	if (!vf) {
+		auto errnum {errno};
+		return merror::Error(
+			generic_category().default_error_condition(errnum),
+			"Error reading the version number from the version file");
+	}
+
+	auto exp_version_number {common::StringToInt(version)};
+	if (!exp_version_number) {
+		return executor::error::MakeError(
+			executor::error::VersionFileError,
+			"Error reading the version number from the version file. Is the file in the correct format? error: "
+				+ exp_version_number.error().String());
+	}
+
+	if (exp_version_number.value() != expected_state_script_version) {
+		return executor::error::MakeError(
+			executor::error::VersionFileWrongVersionError,
+			"Unexpected Artifact script version found: " + to_string(exp_version_number.value()));
+	}
+	return merror::NoError;
+}
+
+bool isValidStateScript(const string &file, State state, Action action) {
+	string expression {
+		"(" + state_map.at(state) + ")" + "_(" + action_map.at(action) + ")_[0-9][0-9](_\\S+)?"};
+	log::Debug(
+		"verifying the State script format of the file: " + file
+		+ " using the regular expression: " + expression);
+	const regex artifact_script_regexp {expression, std::regex_constants::ECMAScript};
+	return regex_match(path::BaseName(file), artifact_script_regexp);
+}
+
+function<bool(const string &)> Matcher(State state, Action action) {
+	return [state, action](const string &file) {
+		return path::IsExecutable(file) and isValidStateScript(file, state, action);
+	};
+}
+
+bool isArtifactScript(State state) {
+	switch (state) {
+	case State::Idle:
+	case State::Sync:
+	case State::Download:
+		return false;
+	default:
```

Comment:
I know I'm being annoying 🙂, but if we spell out the remaining script names, then the compiler will warn us if we ever add any new script types and this table needs updating.

Reply:
No annoying at all!

Good point